src/gpu/d3d11/SDL_gpu_d3d11.c

```diff
@@ -2585,20 +2772,105 @@ static SDL_GpuBuffer* D3D11_CreateGpuBuffer(
 	}
 
 	container = SDL_malloc(sizeof(D3D11BufferContainer));
-	container->usage = usageFlags;
 	container->activeBuffer = buffer;
 	container->bufferCapacity = 1;
 	container->bufferCount = 1;
 	container->buffers = SDL_malloc(
 		container->bufferCapacity * sizeof(D3D11Buffer*)
 	);
 	container->buffers[0] = container->activeBuffer;
+    container->bufferDesc = bufferDesc;
 	container->debugName = NULL;
 
 	return (SDL_GpuBuffer*) container;
 }
 
-static void D3D11_INTERNAL_CycleActiveGpuBuffer(
+static D3D11Buffer* D3D11_INTERNAL_CreateUniformBuffer(
+	D3D11Renderer *renderer,
+    Uint32 sizeInBytes
+) {
+	D3D11_BUFFER_DESC bufferDesc;
+	ID3D11Buffer *bufferHandle;
+	D3D11Buffer *d3d11Buffer;
+	HRESULT res;
+
+	bufferDesc.BindFlags = D3D11_BIND_CONSTANT_BUFFER;
+	bufferDesc.ByteWidth = sizeInBytes;
+	bufferDesc.CPUAccessFlags = D3D11_CPU_ACCESS_WRITE;
+	bufferDesc.MiscFlags = 0;
+	bufferDesc.StructureByteStride = 0;
+	bufferDesc.Usage = D3D11_USAGE_DYNAMIC;
+
+	res = ID3D11Device_CreateBuffer(
+		renderer->device,
+		&bufferDesc,
+		NULL,
+		&bufferHandle
+	);
+	ERROR_CHECK_RETURN("Failed to create uniform buffer", 0);
+
+    d3d11Buffer = SDL_malloc(sizeof(D3D11Buffer));
+    d3d11Buffer->handle = bufferHandle;
+    d3d11Buffer->size = sizeInBytes;
+    d3d11Buffer->srv = NULL;
+    d3d11Buffer->uav = NULL;
+    SDL_AtomicSet(&d3d11Buffer->referenceCount, 0);
+
+	return d3d11Buffer;
+}
+
+static SDL_GpuUniformBuffer *D3D11_CreateUniformBuffer(
+    SDL_GpuRenderer *driverData,
+    Uint32 sizeInBytes
+) {
+    D3D11Renderer *renderer = (D3D11Renderer*) driverData;
+    D3D11UniformBuffer *uniformBuffer;
+	D3D11BufferContainer *container;
+	D3D11Buffer *buffer;
+    D3D11_BUFFER_DESC bufferDesc;
+
+    bufferDesc.BindFlags = D3D11_BIND_CONSTANT_BUFFER;
+    bufferDesc.ByteWidth = sizeInBytes;
+    bufferDesc.CPUAccessFlags = D3D11_CPU_ACCESS_WRITE;
+    bufferDesc.MiscFlags = 0;
+    bufferDesc.StructureByteStride = 0;
+    bufferDesc.Usage = D3D11_USAGE_DYNAMIC;
+
+	buffer = D3D11_INTERNAL_CreateBuffer(
+        renderer,
+        &bufferDesc,
+        sizeInBytes
+    );
+
+    buffer = D3D11_INTERNAL_CreateUniformBuffer(renderer, sizeInBytes);
```

Comment:
is it intended that you D3D11_INTERNAL_CreateBuffer into buffer and then do CreateUniformBuffer and destroy the previous value by overwriting it? Should you be storing the UB into uniformBuffer instead?